File: src/main/java/org/simplemc/simplereserve/SimpleReserve.java
package org.simplemc.simplereserve;

import org.bukkit.ChatColor;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.plugin.java.JavaPlugin;

/**
 * Simple reserve slot plugin the Bukkit API
 *
 * @author Taylor Becker
 */
public class SimpleReserve extends JavaPlugin
{
    /**
     * set up the block listener and Permissions on enable
     */
    public void onEnable()
    {
        loadConfig();

        getLogger().info(getDescription().getName() + " version " + getDescription().getVersion() + " enabled!");
    }

    /**
     * Load the config options from config
     */
    private void loadConfig()
    {
        // ensure config file/options valid
        validateConfig();

        // load
        ConfigurationSection config = getConfig().getConfigurationSection("reserve"); // our config
        ReserveType reserveMethod; // type of reservations

        try
        {
            reserveMethod = ReserveType.valueOf(config.getString("type").toUpperCase());
        }
        // config not set right(enum can't be valueOf'd)
        catch (IllegalArgumentException iae)
        {
            // config not set right, default to both
            config.set("type", "both");
            saveConfig();
            getLogger().info(getDescription().getName() + " config file updated, please check settings!");

            reserveMethod = ReserveType.valueOf(config.getString("type").toUpperCase());
        }

        new SimpleReserveListener(reserveMethod,
                config.getInt("full.cap", 5),
                config.getBoolean("full.reverttokick", false),
                config.getString("kick-message", "Kicked to make room for reserved user!"),
                config.getString("full-message", "The server is full!"),
                config.getString("reserve-full-message", "All reserve slots are full!"),
                this);
    }

    @Override
    public boolean onCommand(CommandSender sender, Command command, String label, String[] args)
    {
        if (command.getName().equalsIgnoreCase("simplereserve"))
        {
            if (sender.hasPermission("simplereserve"))
            {
                // reload command
                if (args.length > 0 && (args[0].equalsIgnoreCase("reload") || args[0].equalsIgnoreCase("r")))
                {
                    if (sender.hasPermission("simplereserve.reload"))
                    {
                        reloadConfig(); // reload file
                        loadConfig(); // read config

                        getLogger().fine("Config reloaded.");
                        sender.sendMessage("SimpleReserve config reloaded");
                    }
                    else
                    {
                        sender.sendMessage(ChatColor.RED + "You do not have permission to do that!");
                    }
                }
                // help command
                else
                {
                    sender.sendMessage(ChatColor.AQUA + "/" + getCommand("simplereserve").getName() + ChatColor.WHITE
                            + " | " + ChatColor.BLUE
                            + getCommand("simplereserve").getDescription());
                    sender.sendMessage("Usage: " + ChatColor.GRAY + getCommand("simplereserve").getUsage());
                }
            }
            else
            {
                sender.sendMessage(ChatColor.RED + "You do not have permission to do that!");
            }
            return true;
        }

        return false;
    }

    /**
     * Validate nodes, if they don't exist or are wrong, set them and resave config
     * <p>
     * Unfortunately we cannot use defaults because there is no reasonable way of copying the defaults
     * into the live config file while maintaining an updated header. We could use saveDefaultConfig
     * but this doesn't help us when changing the config options and a config file already exists and
     * copyDefaults destroys the header.
     * </p>
     */
    private void validateConfig()
    {
        boolean updated = false;
        FileConfiguration config = getConfig(); // our config

        // settings
        if (!config.contains("reserve.type"))
        {
            config.set("reserve.type", "both");
            updated = true;
        }

        if (!config.contains("reserve.full.cap"))
        {
            config.set("reserve.full.cap", 5);
            updated = true;
        }

        if (!config.contains("reserve.full.reverttokick"))
        {
            config.set("reserve.full.reverttokick", false);
            updated = true;
        }

        if (!config.contains("reserve.kick-message"))
        {
            config.set("reserve.kick-message", "Kicked to make room for reserved user!");
            updated = true;
        }

        if (!config.contains("reserve.full-message"))
        {
            config.set("reserve.full-message", "The server is full!");
            updated = true;
        }

        if (!config.contains("reserve.reserve-full-message"))
        {
            config.set("reserve.reserve-full-message", "All reserve slots full!");
            updated = true;
        }

        // if nodes have been updated, update header then save
        if (updated)
        {
            // set header for information
            config
                    .options()
                    .header(
                            "Config nodes:\n"
                                    +
                                    "\n"
                                    +
                                    "reserve.type(enum/string): Type of reserve slots, options:\n"
                                    +
                                    "    full,kick,both,none\n"
                                    +
                                    "reserve.full.cap(int): Max players allowed over capacity if using 'full' method, 0 for no max\n"
                                    +
                                    "reserve.full.reverttokick(boolean): Should we fall back to kick method if we reach max over capacity using full?\n"
                                    +
                                    "kick-message(string): Message player will recieve when kicked to let reserve in\n"
                                    +
                                    "full-message(string): Message player will recieve when unable to join full server\n"
                                    +
                                    "reserve-full-message(string): Message player with reserve privileges will recieve when all reserve slots are full\n"
                                    +
                                    "\n" +
                                    "Reserve Types Overview:\n" +
                                    "-----------------------\n" +
                                    "\n" +
                                    "Full: Allow reserves to log on past the server limit\n" +
                                    "Kick: Attempt to kick a player without kick immunity to make room\n" +
                                    "Both: Both methods of reservation based on Permission\n" +
                                    "    NOTE: If a player has permission for kick and full, full applies\n" +
                                    "None: No reservation. Effectively disables mod without needing to remove\n" +
                                    "");

            // save
            saveConfig();
            getLogger().info(getDescription().getName() + " config file updated, please check settings!");
        }
    }

    /**
     * plugin disabled
     */
    public void onDisable()
    {
        System.out.println("SimpleReserve disabled!");
    }
}

File: src/main/kotlin/org/simplemc/simplereserve/SimpleReserve.kt
package org.simplemc.simplereserve

import org.bukkit.command.Command
import org.bukkit.command.CommandSender
import org.bukkit.plugin.java.JavaPlugin
import org.simplemc.simplereserve.ReserveConfig.FullMethodConfig
import org.simplemc.simplereserve.ReserveConfig.KickMethodConfig
import org.simplemc.simplereserve.ReserveConfig.ReserveMethod

/**
 * Simple reserve slot plugin the Bukkit API
 *
 * @author Taylor Becker
 */
class SimpleReserve : JavaPlugin() {

    lateinit var listener: ReserveLoginListener
    override fun onEnable() {
        listener = ReserveLoginListener(this, loadConfig())

        checkNotNull(getCommand("simplereservereload")).setExecutor(::reloadConfigCommand)

        logger.info { "${description.name} version ${description.version} enabled!" }
    }

    /**
     * Load the config options from config
     */
    private fun loadConfig(): ReserveConfig {
        // ensure config file/options valid
        validateConfig()

        // load
        val config = config.getConfigurationSection("reserve") // our config

        val reserveMethod: ReserveMethod = try {
            ReserveMethod.valueOf(config!!.getString("method")!!.uppercase())
        } catch (_: IllegalArgumentException) {
            // invalid method, default to both
            // config not set right, default to both
            config!!.set("method", "both")
            saveConfig()
            logger.info { "${description.name} config file updated, please check settings!" }

            ReserveMethod.valueOf(config.getString("method")!!.uppercase())
        }

        return ReserveConfig(
            method = reserveMethod,
            serverFullMessage = config.getString("server-full-message", "The server is full!")!!,
            full = FullMethodConfig(
                capacity = config.getInt("full.cap", 5),
                kickFallback = config.getBoolean("full.kick-fallback", false),
                overCapacityMessage = config.getString("full.over-capacity-message", "All reserve slots are full!")!!,
            ),
            kick = KickMethodConfig(
                message = config.getString(
                    "kick.message",
                    "Kicked to make room for reserved user!",
                )!!,
            ),
        )
    }

    private fun reloadConfigCommand(
        sender: CommandSender,
        command: Command,
        label: String,
        vararg args: String,
    ): Boolean {
        listener.close()
        reloadConfig() // reload config file from disk
        listener = ReserveLoginListener(this, loadConfig())

        logger.fine { "Config reloaded." }
        sender.sendMessage("SimpleReserve config reloaded")
        return true
    }

    /**
     * Validate nodes, if they don't exist or are wrong, set them and resave config
     *
     * Unfortunately, we cannot use defaults because there is no reasonable way of copying the defaults into the live
     * config file while maintaining an updated header. We could use [saveDefaultConfig] but this doesn't help us when
     * changing available config options when a config file already exists.
     */
    private fun validateConfig() {
        var updated = false

        if (!config.contains("reserve.method")) {
            config.set("reserve.method", "both")
            updated = true
        }

        if (!config.contains("reserve.server-full-message")) {
            config.set("reserve.server-full-message", "The server is full!")
            updated = true
        }

        if (!config.contains("reserve.full.cap")) {
            config.set("reserve.full.cap", 5)
            updated = true
        }

        if (!config.contains("reserve.full.kick-fallback")) {
            config.set("reserve.full.kick-fallback", false)
            updated = true
        }

        if (!config.contains("reserve.full.over-capacity-message")) {
            config.set("reserve.full.over-capacity-message", "All reserve slots full!")
            updated = true
        }

        if (!config.contains("reserve.kick.message")) {
            config.set("reserve.kick.message", "Kicked to make room for reserved user!")
            updated = true
        }

        // if nodes have been updated, update header then save
        if (updated) {
            // set header for information
            config.options().setHeader(
                listOf(
                    "Config nodes:",
                    "",
                    "reserve.method(enum/string): Method to use for allowing reservees in, options:",
                    "    full,kick,both,none",
                    "reserve.server-full-message(string): Message player will receive when unable to join full server",
                    "reserve.full.cap(int): Max players allowed over capacity if using 'full' method, 0 for no max",
                    "reserve.full.kick-fallback(boolean): Should we fall back to kick method if we reach max over capacity using full?",
                    "reserve.full.over-capacity-message(string): Message player with reserve privileges will receive when all reserve slots are full",
                    "reserve.kick.message(string): Message player will receive when kicked to let reserve in",
                    "",
                    "Reserve Methods Overview:",
                    "-----------------------",
                    "",
                    "Full: Allow reserves to log on past the server limit",
                    "Kick: Attempt to kick a player without kick immunity to make room",
                    "Both: Both methods of reservation based on Permission",
                    "    NOTE: If a player has permission for kick and full, full takes precedence",
                    "None: No reservation. Effectively disables mod without needing to remove",
                    "",
                ),
            )

            // save
            saveConfig()
            logger.info { "${description.name} config file updated, please check settings!" }
        }
    }

    /**
     * Plugin disabled
     */
    override fun onDisable() {
        listener.close()
        println("SimpleReserve disabled!")
    }
}
